agent/metrics/metrics.go

```diff
@@ -0,0 +1,246 @@
+		if err != nil {
+			return nil, err
+		}
+		resourceLabels = res
+	} else {
+		err := fmt.Errorf("Failed to create metric handler: unknown resource type %v", resourceType)
+		return nil, err
+	}
+
+	startTime = time.Now()
+	codeCount = make(map[string]int64)
+
+	return &MetricHandler{
+		projectID:      projectID,
+		metricDomain:   metricDomain,
+		resourceType:   resourceType,
+		resourceLabels: resourceLabels,
+		ctx:            ctx,
+		client:         client,
+	}, nil
+}
+
+func parseGCEResourceLabels(resourceKeyValues string) (*map[string]string, error) {
+	flags := map[string]string{
+		"instance-id":   "",
+		"instance-zone": "",
+	}
+	res := map[string]string{
+		"instance_id": "",
+		"zone":        "",
+	}
+	err := parseResourceLabels(resourceKeyValues, &flags)
+	for key, _ := range res {
+		flagKey := responseCountResourceKeyToFlagName[key]
+		flagValue := flags[flagKey]
+		if flagValue == "" {
+			err := fmt.Errorf("Failed to create metric handler: missing gce_instance resource label (%v)", key)
+			return &res, err
+		}
+		res[key] = flagValue
+	}
+	return &res, err
+}
+
+func parseResourceLabels(resourceKeyValues string, labels *map[string]string) error {
+	if resourceKeyValues == "" {
+		return nil
+	}
+	pairs := strings.Split(resourceKeyValues, ",")
+	for _, p := range pairs {
+		pair := strings.Split(p, "=")
+		if len(pair) != 2 {
+			err := fmt.Errorf("Error parsing monitoringKeyValue('%v'): got %v expressions, wanted 2", p, len(pair))
+			return err
+		}
+		key, value := pair[0], pair[1]
+		if _, ok := (*labels)[key]; ok {
+			(*labels)[key] = value
+		}
+	}
+	return nil
+}
+
+func (h *MetricHandler) GetResponseCountMetricType() string {
+	if h == nil {
+		return ""
+	}
+	return fmt.Sprintf("%s/instance/proxy_agent/response_count", h.metricDomain)
+}
+
+// WriteResponseCodeMetric will gather response codes and write to cloud monarch once sample period is over
+func (h *MetricHandler) WriteResponseCodeMetric(statusCode int) error {
+	if h == nil {
+		return nil
+	}
+	responseCode := fmt.Sprintf("%v", statusCode)
+
+	// Lock out other goroutines
+	h.mu.Lock()
+
+	// Update response code count for the current sample period
+	codeCount[responseCode]++
+
+	// Only write time series after sample period is over
+	if time.Since(startTime) < samplePeriod {
+		h.mu.Unlock()
+		return nil
+	}
+
+	// Write metric
+	log.Printf("WriteResponseCodeMetric|attempting to write metrics at time: %v\n", time.Now())
+	for responseCode, count := range codeCount {
+		responseClass := fmt.Sprintf("%sXX", responseCode[0:1])
+		metricLabels := map[string]string{
+			"response_code":       responseCode,
+			"response_code_class": responseClass,
+		}
+		timeSeries := h.newTimeSeries(h.GetResponseCountMetricType(), metricLabels, newDataPoint(count))
+
+		log.Printf("got %v occurances of %s response code\n", count, responseCode)
+		log.Printf("%v\n\n", timeSeries)
+
+		if err := h.client.CreateTimeSeries(h.ctx, &monitoringpb.CreateTimeSeriesRequest{
```

Comment:
This is a call to an external resource, so we don't want to run it on the same goroutine that's processing a user request.

My suggestions is to have a background goroutine that spins in a loop, pushing updates to stackdriver every $interval seconds. The method to record the metric for a given request just updates the map of codes, and the background goroutine that writes the metrics zeroes out the map when it updates. 
We only lock the map updates (incrementing and zeroing), not the actual call to Stackdriver.

This is similar to what we do for updating the VMID header in the background.  https://github.com/google/inverting-proxy/blob/master/agent/utils/utils.go#L199-L213

Reply:
Done.